### PR TITLE
feat(ir): more flexible dereferencing support for join right hand side

### PR DESCRIPTION
**Pull request by kszucs**
Enables to use fields from parent tables of the join right hand side instead of enforcing to use the same exact table:

```py
t1 = ibis.table(name="t1", schema={"a": "int64", "b": "string"})
t2 = ibis.table(name="t2", schema={"c": "int64", "d": "string"})

t3 = t2.mutate(e=t2.c + 1)
joined = t1.join(t3, [t1.a == t2.c])  # here we use t2.c instead of t3.c
```

Identify ambiguous cases and raise an error, like the following case:

```py
t.join(t, [t.a == t.a])
```

depends on:
- #9043 
- #9041 

fixes https://github.com/ibis-project/ibis/issues/8581

### CODE REVIEW

File: ibis/expr/format.py [hunk 1]
```diff
@@ -151,7 +152,8 @@ def get_defining_frame(expr):
     """Locate the outermost frame where `expr` is defined."""
     for frame_info in inspect.stack()[::-1]:
         for var in frame_info.frame.f_locals.values():
-            if isinstance(var, ir.Expr) and expr.equals(var):
+            # if isinstance(var, ir.Expr) and expr.equals(var):
```
Comment (kszucs) [inline ibis/expr/format.py:155]: Going to apply this change in a different PR.

File: ibis/expr/rewrites.py [hunk 2]
```diff
@@ -23,6 +28,55 @@
...
+            value = value.rel.values.get(value.name)
+            distance += 1
+        if value is not None:
+            yield value, distance
+
+    def dereference(self, value):
+        ambigs = value.find(lambda x: x in self.ambigs, filter=ops.Value)
+        if ambigs:
+            raise IbisInputError(
+                f"Ambiguous field reference {ambigs!r} in expression {value!r}"
```
Comment (kszucs) [inline ibis/expr/rewrites.py:75]: The repr must be improved.

File: ibis/common/graph.py [hunk 3]
```diff
@@ -66,8 +66,9 @@ def _flatten_collections(node: Any) -> Iterator[N]:
             yield item
         elif isinstance(item, (tuple, list)):
             yield from _flatten_collections(item)
-        elif isinstance(item, (dict, frozendict)):
-            yield from _flatten_collections(item.values())
+        elif isinstance(item, dict):
+            items = itertools.chain.from_iterable(item.items())
```
Comment (kszucs) [inline ibis/common/graph.py:70]: Going to pull this out to a separate PR.

File: ibis/expr/types/joins.py [hunk 4]
```diff
@@ -112,47 +112,18 @@ def disambiguate_fields(
...
-def dereference_value(pred, deref_left, deref_right):
-    deref_both = {**deref_left, **deref_right}
-    if isinstance(pred, ops.Comparison) and pred.left.relations == pred.right.relations:
-        left, right = dereference_sides(pred.left, pred.right, deref_left, deref_right)
-        return pred.copy(left=left, right=right)
-    else:
-        return pred.replace(deref_both, filter=ops.Value)
+def dereference_value(pred, deref_left, deref_right, deref_both):
+    return deref_both.dereference(pred)
+    # if isinstance(pred, ops.Comparison) and pred.left.relations == pred.right.relations:
```
Comment (kszucs) [inline ibis/expr/types/joins.py:117]: This code bit is responsible for allowing `a.join(a, [a.col == a.col])`

Comment (kszucs) [reply ibis/expr/types/joins.py:117]: Removing it once I have confirmation about that we are disallowing it.

Comment (cpcloud) [reply ibis/expr/types/joins.py:117]: This seems very ambiguous to me and _somewhat_ unlikely. +1 for removing it

cc @jcrist @gforsyth for opinions

File: ibis/common/graph.py (hunk 3 above)
Comment (kszucs) [reply ibis/common/graph.py:70]: https://github.com/ibis-project/ibis/pull/9041

File: ibis/expr/types/joins.py (hunk 4 above)
Comment (gforsyth) [reply ibis/expr/types/joins.py:117]: I am +1 on removing it

File: ibis/expr/format.py (hunk 1 above)
Comment (kszucs) [reply ibis/expr/format.py:155]: https://github.com/ibis-project/ibis/pull/9043

File: ibis/expr/types/joins.py (hunk 4 above)
Comment (jcrist) [reply ibis/expr/types/joins.py]: Also +1. Rip it out.

File: ibis/expr/rewrites.py (hunk 2 above)
Comment (kszucs) [reply ibis/expr/rewrites.py:75]: Deferring it to a follow-up.

File: ibis/backends/impala/tests/snapshots/test_sql/test_nested_join_base/out.sql [hunk 5]
```diff
@@ -7,7 +7,7 @@ WITH `t1` AS (
     1
 )
 SELECT
-  `t5`.`uuid`,
+  `t3`.`uuid`,
```
Comment (cpcloud) [inline ibis/backends/impala/tests/snapshots/test_sql/test_nested_join_base/out.sql:10]: This looks incorrect to me. Check out the expression:


```python
    t = ibis.table(dict(uuid="string", ts="timestamp"), name="t")
    counts = t.group_by("uuid").size()
    max_counts = counts.group_by("uuid").aggregate(max_count=lambda x: x[1].max())
    result = max_counts.left_join(counts, "uuid").select(counts)
```

The `uuid` column in the final `.select` call is coming from the **right** side of the join, so it should remain `t5.uuid` I think.

Comment (cpcloud) [reply ibis/backends/impala/tests/snapshots/test_sql/test_nested_join_base/out.sql:10]: The right side `Field` cannot dereference to the left side field because this is a left outer join, so the join key values for a given row may be different.

Comment (kszucs) [reply ibis/backends/impala/tests/snapshots/test_sql/test_nested_join_base/out.sql:10]: The following is the repr (after setting `ibis.options.repr.show_variables = True`):

```
t := UnboundTable: t
  uuid string
  ts   timestamp

counts := Aggregate[t]
  groups:
    uuid: t.uuid
  metrics:
    CountStar(t): CountStar(t)

max_counts := Aggregate[counts]
  groups:
    uuid: counts.uuid
  metrics:
    max_count: Max(counts['CountStar(t)'])

JoinChain[max_counts]
  JoinLink[left, counts]
    max_counts.uuid == counts.uuid
  values:
    uuid:         counts.uuid
    CountStar(t): counts['CountStar(t)']
```

Isn't this the expected expr?

Comment (cpcloud) [reply ibis/backends/impala/tests/snapshots/test_sql/test_nested_join_base/out.sql:10]: You're right! I misread. Resolved.

File: ibis/expr/rewrites.py [hunk 6]
```diff
@@ -23,6 +28,55 @@
 name = var("name")
 
 
+class DerefMap(Concrete, Traversable):
```
Comment (cpcloud) [inline ibis/expr/rewrites.py:31]: Can you add a docstring? This is a pretty important part of the library, and it would help folks understand how it fits into everything if there were some docs.

File: ibis/expr/tests/test_newrels.py [hunk 7]
```diff
@@ -1233,9 +1299,11 @@ def test_self_join_with_intermediate_selection():
...
         assert join.op() == expected
 
-    aa = a.join(a, [a.a == a.a])["a", "b_right"]
-    aaa = aa.join(a, [aa.a == a.a])
-    with join_tables(a, a, a) as (r1, r2, r3):
+    a_ = a.view()
+    a__ = a.view()
+    aa = a.join(a_, [a.a == a_.a])["a", "b_right"]
+    aaa = aa.join(a__, [aa.a == a__.a])
+    with join_tables(aaa) as (r1, r2, r3):
```
Comment (cpcloud) [inline ibis/expr/tests/test_newrels.py:1306]: Hm, let's come up with some more unique variable names here 😂

File: ibis/expr/tests/test_newrels.py [hunk 8]
```diff
@@ -1173,10 +1205,28 @@ def test_join_chain_gets_reused_and_continued_after_a_select():
 def test_self_join_extensive():
     a = ibis.table(name="a", schema={"a": "int64", "b": "string"})
 
-    aa = a.join(a, [a.a == a.a])
+    with pytest.raises(IbisInputError, match="Ambiguous field reference"):
```
Comment (cpcloud) [inline ibis/expr/tests/test_newrels.py:1208]: Nice, glad we can detect and raise for this case!

**Review by cpcloud — APPROVED** [review]
LGTM mostly, just a few small requests.

Would really like to see a docstring for `DerefMap` but I won't block on that.

File: ibis/expr/tests/test_newrels.py (hunk 7 above)
Comment (kszucs) [reply ibis/expr/tests/test_newrels.py:1306]: I can do a1 and a2 :)

File: ibis/common/typing.py [hunk 9]
```diff
@@ -259,3 +260,21 @@ class Coercible(Abstract):
...
     def __coerce__(cls, value: Any, **kwargs: Any) -> Self: ...
+
+
+def get_defining_frame(obj):
+    """Locate the outermost frame where `obj` is defined."""
+    for frame_info in inspect.stack()[::-1]:
+        for var in frame_info.frame.f_locals.values():
+            if obj is var:
+                return frame_info.frame
+    raise ValueError(f"No defining frame found for {obj}")
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/common/typing.py]: I don't think it is important for now since it is an opt-in feature.

File: ibis/expr/tests/test_newrels.py (hunk 7 above)
Comment (cpcloud) [reply ibis/expr/tests/test_newrels.py:1306]: SGTM, thanks!
